x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -1022,10 +1024,15 @@ private AbstractRunnable fillGapRunnable(
                     streamFactory,
                     start,
                     Math.toIntExact(gap.end() - start),
-                    progress -> gap.onProgress(start + progress)
+                    progress -> gap.onProgress(start + progress),
+                    l.<Void>map(unused -> {
+                        assert regionOwners.get(ioRef) == CacheFileRegion.this;
+                        assert CacheFileRegion.this.hasReferences() : CacheFileRegion.this;
```

Comment:
I think I prefer to keep them, since there is now async processing in between. In case the (stack) caller for instance frees the ref-count wrongly, then this could catch that (when the processing is really async).